libvast/builtins/operators/enumerate.cpp

```diff
@@ -0,0 +1,143 @@
+
+#include <unordered_set>
+
+namespace vast::plugins::enumerate {
+
+namespace {
+
+// TODO: refactor this once we have a modifier that allows for toggling
+// between intra-schema and inter-schema processing.
+class enumerate_operator final : public crtp_operator<enumerate_operator> {
+public:
+  explicit enumerate_operator(std::string field) : field_{std::move(field)} {
+    if (field_.empty())
+      field_ = "#";
+  }
+
+  auto
+  operator()(generator<table_slice> input, operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    // Per-schema state.
+    auto current_type = type{};
+    std::unordered_map<type, uint64_t> offsets;
+    // Create transformation to prepend  column to slice.
+    auto transformations = std::vector<indexed_transformation>{};
+    auto function = [&](struct record_type::field field,
+                        std::shared_ptr<arrow::Array> array) mutable
+      -> std::vector<
+        std::pair<struct record_type::field, std::shared_ptr<arrow::Array>>> {
+      // Create a column with the RIDs.
+      auto n = array->length();
+      auto rid_type = uint64_type{};
+      auto builder
+        = uint64_type::make_arrow_builder(arrow::default_memory_pool());
+      auto reserve_result = builder->Reserve(n);
+      VAST_ASSERT_CHEAP(reserve_result.ok(), reserve_result.ToString().c_str());
+      // Fill the column.
+      auto& offset = offsets[current_type];
+      for (uint64_t i = 0; i < detail::narrow_cast<uint64_t>(n); ++i) {
+        auto append_result
+          = append_builder(rid_type, *builder, view<uint64_t>{offset + i});
+        VAST_ASSERT(append_result.ok(), append_result.ToString().c_str());
+      }
+      offset += n;
+      // Replace first column with a pair of (RID, first).
+      auto rid_array = builder->Finish().ValueOrDie();
+      return {
+        {{field_, rid_type}, rid_array},
+        {field, array},
+      };
+    };
+    transformations.emplace_back(offset{0}, std::move(function));
+    for (auto&& slice : input) {
+      if (slice.rows() == 0) {
+        co_yield {};
+      } else if (skipped_schemas_.contains(slice.schema())) {
+        co_yield slice;
+      } else if (caf::get<record_type>(slice.schema())
+                   .resolve_key(field_)
+                   .has_value()) {
+        ctrl.warn(caf::make_error(ec::unspecified,
+                                  fmt::format("ignoring schema {} with already "
+                                              "existing enumeration key {}",
+                                              slice.schema().name(), field_)));
+        skipped_schemas_.insert(slice.schema());
+        co_yield slice;
+      } else {
+        current_type = slice.schema();
+        co_yield transform_columns(slice, transformations);
+      }
+    }
+  }
+
+  auto to_string() const -> std::string override {
+    return "enumerate";
+  }
+
+private:
+  std::string field_;
+  mutable std::unordered_set<type> skipped_schemas_;
```

Comment:
The call operator is const for a good reason, what you're doing here may lead to a crash. Just keep the variable in the generator coroutine stack.

Reply:
Check. (f07db642f4)